consensus/istanbul/backend/engine.go

```diff
@@ -272,29 +270,103 @@ func (sb *Backend) verifyCommittedSeals(chain consensus.ChainReader, header *typ
 		return errEmptyCommittedSeals
 	}
 
+	// Check the signatures on the current header
+	snap, err := sb.snapshot(chain, header.Number.Uint64()-1, header.ParentHash, parents)
+	if err != nil {
+		return err
+	}
 	validators := snap.ValSet.Copy()
-	// Check whether the committed seals are generated by parent's validators
+	err = sb.checkValidatorSignatures(header.Hash(), validators, extra.Bitmap, extra.CommittedSeal)
+	if err != nil {
+		return err
+	}
+
+	// Check parent signatures only after block 1.
+	// The genesis block is skipped since it has no parents
+	// The first block is also skipped since its parent
+	// is the genesis block which contains no parent signatures
+	if number > 1 {
+		sb.logger.Debug("verifyCommittedSeals: verifying parent block for block", "num", number)
+		var parentValidators istanbul.ValidatorSet
+		// parents of epoch blocks have different validator sets since they
+		// belong to the previous epoch.
+		if number%sb.config.Epoch == 0 {
+			sb.logger.Debug("verifyCommittedSeals: it's an epoch block!")
+			var parent *types.Header
+			var cache []*types.Header
+			if len(parents) > 1 {
+				cache = parents[:len(parents)-1]
+				parent = parents[len(parents)-1]
+				sb.logger.Debug("verifyCommittedSeals: was given a cache", "cache-len", len(cache), "parent-num-from-cache", parent.Number.Uint64())
+			} else {
+				cache = parents
+				parent = chain.GetHeaderByNumber(number - sb.config.Epoch)
```

Comment:
I think we can probably get away with not fetching the parent anywhere in this function and just using the block number for `sb.snapshot`. Would clean a lot of this code up